cloudimg/aws.py

```diff
@@ -643,3 +703,92 @@ def tag_image(self, image, tags):
         res = image.create_tags(**attrs)
         log.debug("Tag image \"%s\" results: %s" % image.name, res)
         return res
+
+    def deregister_image(self, image):
+        """
+        Deregisters AMI image from AWS.
+
+        Args:
+            image (Image): An EC2 Image
+
+        Returns:
+            Id of deregistered image. (str)
+        """
+        image_id, image_name = image.id, image.name
+        logging.info("Deregistering image %s (%s)", image_id, image_name)
+
+        res = image.deregister()
+
+        logging.debug("Deregister image %s (%s) results: %s",
+                      image_id, image_name, res)
+
+        return image_id
+
+    def delete_snapshot(self, snapshot):
+        """
+        Deletes snapshot from AWS.
+
+        Args:
+            snapshot (Snapshot): An EC2 Snapshot
+
+        Returns:
+            Id of deleted snapshot. (str)
+        """
+        snapshot_id = snapshot.id
+        logging.info("Deleting snapshot %s", snapshot_id)
+
+        res = snapshot.delete()
+
+        logging.debug("Deleted snapshot %s results: %s", snapshot_id, res)
+
+        return snapshot_id
+
+    def delete(self, metadata):
+        """
+        Deletes AMI images and snapshot for given metadata.
+
+        If specified image is not found, we try to find snapshot
+        by other provided metdadata.
+
+        Args:
+            metadata (AWSDeleteMetadata): Metadata about the image
+
+        Returns:
+            tuple (image_id[Str], snapshot_id[Str]]) of removed objects.
+        """
+        deleted_image_id = None
+        deleted_snapshot_id = None
+
+        log.info('Searching for image: %s', metadata.image_id)
+        image = (
+            self.get_image_by_id(metadata.image_id) or
+            self.get_image_by_name(metadata.image_name)
+        )
+
+        if image:
+            # extract snapshot_id from existing image
+            snapshot_id = image.block_device_mappings[0]["Ebs"]["SnapshotId"]
+            snapshot = self.get_snapshot_by_id(snapshot_id)
+
+            deleted_image_id = self.deregister_image(image)
+
+        # image doesn't exist, let's try to find snapshot
+        # by other provided metadata
+        else:
+            log.info('Image does not exist: %s', metadata.image_id)
+            log.info('Searching for snapshot: %s', metadata.snapshot_id)
+            snapshot = (
+                self.get_snapshot_by_id(metadata.snapshot_id) or
+                self.get_snapshot_by_name(metadata.snapshot_name)
+            )
+
+        if metadata.skip_snapshot:
+            log.info("Skipping snapshot deletion because"
+                     "skip_snapshot is set to True")
```

Comment:
This will currently log as `... deletion becauseskip_snapshot is ...`.  Should be:

```suggestion
            log.info("Skipping snapshot deletion because "
                     "skip_snapshot is set to True")
```

Reply:
I'm also wondering, given that we already looked up the snapshot, is it possibly valuable to log the specific snapshot not being deleted?

I guess it could be.  In cases where the snapshot can only be found due to its relationship with the image, then if we accidentally run with skip_snapshot=True when it should have been False, you cannot fix the problem by re-running with skip_snapshot=False (as the image pointing at the snapshot is gone already) and you cannot manually delete the snapshot unless the ID was logged somewhere.